### DISCUSSION TITLE
New Hook - auth.body.email

### DISCUSSION THREAD

**Discussion by kenzik** [directus/directus#16712] [Feature Requests]
I have a need to modify the local driver's payload's `email` value prior to the user retrieval (which is before the existing `auth.login` hook.)

I've made a new hook called `auth.body.email` that is evented prior to the `userLoginSchema.validate()` function in the local auth driver.

This is a very simple change which allows custom computing of the email field prior to (1) validation; (2) user lookup and; (3) local driver authentication. 

This only impacts the JS-SDK/API -- the UI seems to have a separate "must be an email" validator/login schema.

Eg:

```
POST /auth/login
{ 
  "email": "username",
  "password": "123pass"
}
```

May be transformed to:

```
{
  "email": "regionXYZ.some-uuid+username@company.com",
  "password": "123pass"
}
```

Current functioning diff:

```
--- a/api/src/auth/drivers/local.ts
+++ b/api/src/auth/drivers/local.ts
@@ -6,6 +6,8 @@ import { User } from '../../types';
 import { InvalidCredentialsException, InvalidPayloadException } from '../../exceptions';
 import { AuthenticationService } from '../../services';
 import asyncHandler from '../../utils/async-handler';
+import emitter from '../../emitter';
+import getDatabase from '../../database';
 import env from '../../env';
 import { respond } from '../../middleware/respond';
 import { COOKIE_OPTIONS } from '../../constants';
@@ -71,6 +73,28 @@ export function createLocalAuthRouter(provider: string): Router {
                 schema: req.schema,
             });

+            let updatedEmail;
+            if (req.body?.email) {
+                updatedEmail = await emitter.emitFilter(
+                    'auth.body.email',
+                    req.body?.email,
+                    {
+                        status: 'pending',
+                        user: undefined,
+                        provider,
+                    },
+                    {
+                        database: getDatabase(),
+                        schema: null,
+                        accountability: null,
+                    }
+                );
+            }
+
+            if (updatedEmail) {
+                req.body.email = updatedEmail;
+            }
+
             const { error } = userLoginSchema.validate(req.body);

             if (error) {
```

If this is something that would indeed be reviewed and merged in, I'll make a PR with the above and doc updates.

**Comment by kenzik** [comment]
Actually, this can be accomplished a bit deeper and earlier in the code with `init` and just supplying your own middleware.

I'm not sure as to the best approach here. 

The filter I added is very simple -- explicit and single-purposed later in the request cycle and provides the `knex` object; whereas `init` gives you the entire Express app object to use before Directus, which can be much more complicated and dangerous.

For now I'll proceed with my own middleware in `init` since it's officially available.

**Comment by rijkvanzanten** [comment]
Down to discuss a new auth filter hook! I do think it should be a bit broader than just the email field on the local driver though. Maybe there should be a middleware injection point for the "when the login driver route is called"? That way you could intercept this for local, but also add in custom logic for ldap logins or openid/saml/oauth2 redirects 🤔

**Reply by kenzik** [reply]
Wouldn't one of the `init` hooks already suffice? 

I haven't dug into anything but `middlewares.before`, but with its access to the express `app` object, it can check the request object's path to determine the (implied) route already. It's a bit more low-level to get to the end result, of course.

I'm up for whatever seems best here. I started to go down the path of a 'custom' auth driver with a new mapping, but it looked like there was a bit surgery needed to support that (and I wasn't sure as to the roadmap for the core team to support work something like that in -- so many discussions here and there about it.)

Anyway, are you suggesting a "when the login driver route is called **by the local driver**" as the hook point? I can certainly dig into that if that's what you are after.

**Reply by rijkvanzanten** [reply]
Oh yeah you could do anything through `init` really, as (like you said) have access to the full express `app`. That's also the way to immediately get something spun up (without a new core functionality). That being said, I'm down to discuss adding a new filter hook entrypoint in the auth pipeline, but we should figure out a spec that works across drivers (so we don't end up with a mismatch of various events for specific drivers) 🙂

**Reply by kenzik** [reply]
I can dig into that and come up with a proposed plan.

That said, I'm just starting to familiarize myself with the codebase and where things are, so feel free to point me in a certain direction... 

ed: Nevermind... probably the `auth` folder and middleware. :)

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
